### PR TITLE
Prevent Timer's lock with wrong dns

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/1359

I was able to prove that `Peer.OnTimer` could hang the AKKA messages.

POC:

image

We should avoid any expensive procedure in `OnReceive` methods

### CODE REVIEW

**Comment by shargon** [comment]
@lock9 I know that you have a problems with the chain too, please take a look if this can help you

**Comment by erikzhang** [comment]
We can do a DNS query after the node starts and then cache the records. After that we don't need to query every time.

**Comment by shargon** [comment]
I will change it as your advice

File: src/neo/ProtocolSettings.cs [hunk 1]
```diff
@@ -61,18 +63,54 @@ private ProtocolSettings(IConfigurationSection section)
...
-                    "seed1.neo.org:10333",
-                    "seed2.neo.org:10333",
-                    "seed3.neo.org:10333",
-                    "seed4.neo.org:10333",
-                    "seed5.neo.org:10333"
+                    GetIpEndPoint("seed1.neo.org:10333"),
+                    GetIpEndPoint("seed2.neo.org:10333"),
+                    GetIpEndPoint("seed3.neo.org:10333"),
+                    GetIpEndPoint("seed4.neo.org:10333"),
+                    GetIpEndPoint("seed5.neo.org:10333")
```
Comment (erikzhang) [inline src/neo/ProtocolSettings.cs]: This may block the initialization. Moreover, these DNS queries can be performed in parallel.

**Comment by lock9** [comment]
Following Erik's comment, I think we could use the DNS TTL to cache this information if it is not being cached already. Shouldn't the lower network levels do this caching for us already?

**Comment by shargon** [comment]
@lock9 take a look to https://github.com/neo-project/neo/issues/1359 where i explained the problem.

The summary is this, cache the dns at the beginning in order to save time if the dns resolution spend more time than expected.

**Comment by shargon** [comment]
Also we should check if akka has a mechanism for logging messages if they spend more time than expected.

**Comment by lock9** [comment]
Hmm, I'm not very confident this will solve the issue. I understand what you need, but in this case, we need to test it to know. I don't know if the OS is not caching DNS already or if this is going to have the desired outcome.
Did you test it with an invalid domain to verify that it improves the behavior?

**Comment by shargon** [comment]
I tested it and this solve my issue. Maybe `Dns.GetHostEntry` make more things than resolve the name, because it throw an `SocketException` on my machine. 

```powershell
PS C:\Users\XXX> Resolve-DnsName seed1.neo.org | Format-Table -AutoSize
Name          Type TTL Section IPAddress
----          ---- --- ------- ---------
seed1.neo.org A    599 Answer  47.105.64.71


PS C:\Users\XXX> ping seed1.neo.org
Pinging seed1.neo.org [47.105.64.71] with 32 bytes of data:
Reply from 47.105.64.71: bytes=32 time=402ms TTL=45
Reply from 47.105.64.71: bytes=32 time=401ms TTL=45
Request timed out.
Request timed out.

Ping statistics for 47.105.64.71:
    Packets: Sent = 4, Received = 2, Lost = 2 (50% loss),
Approximate round trip times in milli-seconds:
    Minimum = 401ms, Maximum = 402ms, Average = 401ms
```

File: src/neo/ProtocolSettings.cs [hunk 2]
```diff
@@ -61,18 +63,63 @@ private ProtocolSettings(IConfigurationSection section)
                 };
             IConfigurationSection section_sl = section.GetSection("SeedList");
             if (section_sl.Exists())
-                this.SeedList = section_sl.GetChildren().Select(p => p.Get<string>()).ToArray();
+            {
+                this.SeedList = section_sl.GetChildren()
+                      .AsParallel()
+                      .Select(p => GetIpEndPoint(p.Get<string>()))
```
Comment (erikzhang) [inline src/neo/ProtocolSettings.cs]: If the DNS query takes 1 minute, then our program will be blocked for 1 minute. I think we should move this logic to `LocalNode` and then query it asynchronously when the node starts.

File: src/neo/Network/P2P/LocalNode.cs [hunk 3]
```diff
@@ -56,6 +63,17 @@ public LocalNode(NeoSystem system)
...
+
+                SeedList = ProtocolSettings.Default.SeedList
+                    .Select(u => new Seed() { HostAndPort = u })
+                    .ToArray();
+                    
+                // Start dns resolution in parallel
+
+                foreach (var seed in SeedList)
+                {
+                    Task.Run(() => seed).PipeTo(Self, Sender);
```
Comment (erikzhang) [inline src/neo/Network/P2P/LocalNode.cs]: Query DNS records asynchronously here, and pipe the results to `Self`.

File: src/neo/Network/P2P/LocalNode.cs [hunk 4]
```diff
@@ -56,6 +63,22 @@ public LocalNode(NeoSystem system)
...
+
+                foreach (var seed in SeedList)
+                {
+                    Task.Run(() =>
+                    {
+                        seed.EndPoint = GetIpEndPoint(seed.HostAndPort);
+                        return seed;
+                    })
+                    .PipeTo(Self);
+                }
```
Comment (erikzhang) [inline src/neo/Network/P2P/LocalNode.cs]: We can query DNS records in parallel. Use `Task.WhenAll()`.

**Comment by shargon** [comment]
I like the changes

**Pull request merged**